src/SpeedJumpIrm.sol

```diff
@@ -120,43 +122,33 @@ contract AdaptativeCurveIrm is IIrm {
 
         int256 startRateAtTarget = rateAtTarget[id];
 
-        // First interaction.
         if (startRateAtTarget == 0) {
+            // First interaction.
             return (uint256(_curve(INITIAL_RATE_AT_TARGET, err)), INITIAL_RATE_AT_TARGET);
         } else {
             // Note that the speed is assumed constant between two interactions, but in theory it increases because of
             // interests. So the rate will be slightly underestimated.
             int256 speed = ADJUSTMENT_SPEED.wMulDown(err);
-
             // market.lastUpdate != 0 because it is not the first interaction with this market.
             // Safe "unchecked" cast because block.timestamp - market.lastUpdate <= block.timestamp <= type(int256).max.
             int256 elapsed = int256(block.timestamp - market.lastUpdate);
             int256 linearAdaptation = speed * elapsed;
-            int256 adaptationMultiplier = MathLib.wExp(linearAdaptation);
+
             // endRateAtTarget is bounded between MIN_RATE_AT_TARGET and MAX_RATE_AT_TARGET.
             int256 endRateAtTarget =
-                startRateAtTarget.wMulDown(adaptationMultiplier).bound(MIN_RATE_AT_TARGET, MAX_RATE_AT_TARGET);
-            int256 endBorrowRate = _curve(endRateAtTarget, err);
-
-            // Then we compute the average rate over the period.
-            // Note that startBorrowRate is defined in the computations below.
-            // avgBorrowRate = 1 / elapsed * ∫ startBorrowRate * exp(speed * t) dt between 0 and elapsed
-            //               = startBorrowRate * (exp(linearAdaptation) - 1) / linearAdaptation
-            //               = (endBorrowRate - startBorrowRate) / linearAdaptation
-            // And for linearAdaptation around zero: avgBorrowRate ~ startBorrowRate = endBorrowRate.
-            // Also, when it is the first interaction (rateAtTarget = 0).
-            int256 avgBorrowRate;
-            if (linearAdaptation == 0) {
-                avgBorrowRate = endBorrowRate;
-            } else {
-                int256 startBorrowRate = _curve(startRateAtTarget, err);
-                avgBorrowRate = (endBorrowRate - startBorrowRate).wDivDown(linearAdaptation);
+                startRateAtTarget.wMulDown(MathLib.wExp(linearAdaptation)).bound(MIN_RATE_AT_TARGET, MAX_RATE_AT_TARGET);
+
+            // Then we compute the average rate over the period, with a Riemann sum.
+            int256 averageRateAtTarget;
+            int256 step = linearAdaptation / N_STEPS;
+            for (int256 k = 1; k <= N_STEPS; k++) {
+                averageRateAtTarget += startRateAtTarget.wMulDown(MathLib.wExp(step * k)).bound(
```

Comment:
should we multiply `averageRateAtTarget` by `step` at the end?

Reply:
oh right indeed

Reply:
> should we multiply averageRateAtTarget by step at the end?

Before the division by $N_{STEPS}$, this is a sample of the function: $averageRateAtTarget = \sum_k sample(k) / N_ {STEPS}$. Usually we write $averageRateAtTarget = (\sum_k sample(k) * step) / linearVariation$, but it is the same thing as above, simplifying by $linearVariation$ in each of the terms of the sum because $step = linearVariation / N_{STEPS}$